### PR TITLE
[WIP] Share dialog files sidebar

### PR DESCRIPTION
**Pull request by blizzz**
supposed to implement #17665 
- [ ] TODO: modularize share dialog into reusable view and item model
- [ ] TODO: ensure backwards compatibility with showDropDown()
- [ ] TODO: embed share dialog in share tab
- [ ] TODO: write unit tests
- [ ] REQUIRES: https://github.com/owncloud/core/issues/9227

currently this is based on files-rightsidebar

cc  @PVince81

### CODE REVIEW

File: apps/files_sharing/css/sharetabview.css [hunk 1]
```diff
@@ -1,3 +1,72 @@
...
+	white-space: nowrap;
+}
+
+.shareTabView input[type="checkbox"] {
+	margin:0 3px 0 8px;
+	vertical-align: middle;
+}
+
+.shareTabView input[type="text"], .shareTabView input[type="password"] {
+	width: 91%;
```
Comment (PVince81) [inline apps/files_sharing/css/sharetabview.css]: That's some nice precise value you have here :wink:

File: apps/files_sharing/js/sharetabview.js [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
 (function() {
 	var TEMPLATE =
-		'<div>Owner: {{owner}}';
+		'<div><span>Owner: {{owner}}</span>';
```
Comment (PVince81) [inline apps/files_sharing/js/sharetabview.js]: Missing closing `</div>` tag.

File: core/js/sharedialogview.js [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	 */
+	var ShareDialogView = function(id) {
+		this.initialize(id);
+	};
+
+	/**
+	 * @memberof OCA.Sharing
+	 */
+	ShareDialogView.prototype = {
+		/** @member {OC.Share.ShareItemModel} **/
```
Comment (PVince81) [inline core/js/sharedialogview.js]: I usually used `@lends` but if that works then fine

Comment (PVince81) [inline core/js/sharedialogview.js]: Ah wait you need to use `@type OC.Share.ShareItemModel`

File: core/js/sharedialogview.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+		_itemModel: null,
+
+		/** @var {string} **/
```
Comment (PVince81) [inline core/js/sharedialogview.js]: Use `@type`

Comment (PVince81) [inline core/js/sharedialogview.js]: without brackets

Comment (PVince81) [inline core/js/sharedialogview.js]: Dammit I've been doing it wrong, it needs brackets, see http://usejsdoc.org/tags-type.html

File: core/js/sharedialogview.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+			if (!this._templates[key]) {
+				this._templates[key] = Handlebars.compile(template);
+			}
+			return this._templates[key];
+		},
+
+		/**
+		 * returns the info template for remote sharing
+		 *
+		 * @returns {Handlebars}
```
Comment (PVince81) [inline core/js/sharedialogview.js]: I believe the return type is just a plain `Function`, but I might be mistaken

File: core/js/shareitemmodel.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+			return this._reshare.share_with;
+		},
+
+		getReshareType: function() {
+			return this._reshare.share_type;
+		},
+
+		_retrieveData: function() {
+			/** var {OC.Share.Types.ShareItemInfo} **/
+			var data = OC.Share.loadItem(this._itemType, this._itemSource);
```
Comment (PVince81) [inline core/js/shareitemmodel.js]: Please get rid of synchronous calls.

I suggest to do the loading call outside of the model, then inject the values/data in the constructor.

**Comment by PVince81** [comment]
Great progress, it already looks much better than the old dialog code :smile:

File: core/js/share.js [hunk 7]
```diff
@@ -368,6 +374,21 @@ OC.Share={
 		});
 	},
 	showDropDown:function(itemType, itemSource, appendTo, link, possiblePermissions, filename) {
+		var itemModel = new OC.Share.ShareItemModel(itemType, itemSource);
```
Comment (PVince81) [inline core/js/share.js]: Here call OC.Share.loadItem() which is still sync ajax, then inject it into the model.

However in the sidebar it needs to be async, so probably need to rewrite another OC.Share.loadItem

File: core/js/sharedialogview.js (hunk 4 above)
Comment (blizzz) [inline core/js/sharedialogview.js]: Ain't http://usejsdoc.org/tags-property.html even better?

Comment (PVince81) [inline core/js/sharedialogview.js]: Property is only used to add a non-existing property on virtual objects.
Basically properties that exist only in the documentation but there is no code to explicitly define them.
But in our case here we do define the properties in the object/prototype, so we should use `@type` to document their type.

You can try generating the JS docs with ./buildjsdocs.sh then check build/jsdoc/index.html to see how they look like

File: core/js/sharedialogview.js (hunk 5 above)
Comment (blizzz) [inline core/js/sharedialogview.js]: you're right

File: core/js/sharedialogview.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+		_showLink: true,
+
+		/** @type {unknown} **/
+		_possiblePermissions: null,
+
+		/** @type {string} **/
+		tagName: 'div',
+
+		initialize: function() {
+			if(!this.model instanceof OC.Share.ShareItemModel) {
```
Comment (PVince81) [inline core/js/sharedialogview.js]: beware that JS cannot detect subclasses this way

Comment (PVince81) [inline core/js/sharedialogview.js]: (because there are no subclasses in JS, just prototype duplication)

File: core/js/shareitemmodel.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+		getReshareWith: function() {
+			return this._reshare.share_with;
+		},
+
+		getReshareType: function() {
+			return this._reshare.share_type;
+		},
+
+		_retrieveData: function() {
```
Comment (PVince81) [inline core/js/shareitemmodel.js]: This should eventually be the "fetch()" method http://backbonejs.org/#Model-fetch of backbone models (or parse() http://backbonejs.org/#Model-parse to convert the JSON back to the expected structure).

File: core/js/shareitemmodel.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	 *
+	 */
+	var ShareItemModel = function(itemType, itemSource) {
+		this.initialize(itemType, itemSource);
+	};
+
+	/**
+	 * @memberof OCA.Sharing
+	 */
+	ShareItemModel.prototype = {
```
Comment (PVince81) [inline core/js/shareitemmodel.js]: extend from OC.Backbone.Model :smile:

**Comment by PVince81** [comment]
The sidebar was merged, you might want to rebase.

**Comment by blizzz** [comment]
https://github.com/isaacs/github/issues/18 *sigh*

I will rebase and open a PR against master a little later to-:night_with_stars:

File: core/js/sharedialogview.js (hunk 5 above)
Comment (blizzz) [inline core/js/sharedialogview.js]: Fixed

File: core/js/sharedialogview.js (hunk 8 above)
Comment (blizzz) [inline core/js/sharedialogview.js]: Removed

File: core/js/shareitemmodel.js (hunk 10 above)
Comment (blizzz) [inline core/js/shareitemmodel.js]: done

**Comment by blizzz** [comment]
This is not the PR you are looking for. You go here https://github.com/owncloud/core/pull/18185

**Pull request closed without merging**